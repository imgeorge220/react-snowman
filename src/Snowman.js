import React, { useState } from "react";
import "./Snowman.css";
import { randomWord, ENGLISH_WORDS } from "./words";
import img0 from "./0.png";
import img1 from "./1.png";
import img2 from "./2.png";
import img3 from "./3.png";
import img4 from "./4.png";
import img5 from "./5.png";
import img6 from "./6.png";

function Snowman({ maxWrong, images, words }) {
  /** by default, allow 6 guesses and use provided gallows images. */

  const [nWrong, updateNWrong] = useState(0);
  const [guessed, updateGuessed] = useState(new Set());
  const [answer, updateAnswer] = useState(randomWord(words));

  /** guessedWord: show current-state of word:
    if guessed letters are {a,p,e}, show "app_e" for "apple"
  */
  function guessedWord() {
    return answer
      .split("")
      .map(ltr => (guessed.has(ltr) ? ltr : "_"));
  }

  /** handleGuest: handle a guessed letter:
      - add to guessed letters
      - if not in answer, increase number-wrong guesses
  */
  function handleGuess(evt) {
    let ltr = evt.target.value;
    updateGuessed(g => {
      const newGuessed = new Set(g);
      newGuessed.add(ltr);
      return newGuessed;
    });

    updateNWrong(n => n + (answer.includes(ltr) ? 0 : 1));
  }

  /** generateButtons: return array of letter buttons to render */
  function generateButtons() {
    return "abcdefghijklmnopqrstuvwxyz".split("").map(ltr => (
      <button
        key={ltr}
        value={ltr}
        onClick={handleGuess}
        disabled={guessed.has(ltr)}
        data-testid={`btn-${ltr}`}
      >
        {ltr}
      </button>
    ));
  }

  function resetGame() {
    updateNWrong(0);
    updateGuessed(new Set());
    updateAnswer(randomWord(words));
  }

  const buttonArea =
    <div>
      <p>Number of Guesses Left: {maxWrong - nWrong}</p>
      <p className="Snowman-word" data-testid="word">{guessedWord()}</p>
      <p>{generateButtons()}</p>
    </div>;

  const failArea =
    <div>
      <p>GAME OVER!</p>
      <p>Correct word was: {answer}</p>
    </div>;

  /** render: render game */
  return (
    <div className="Snowman">
      <img src={images[nWrong]} data-testid={`img-${images[nWrong]}`} alt="" />
      {nWrong < maxWrong ? buttonArea : failArea}
      <button onClick={resetGame} data-testid="reset-btn">RESET</button>
    </div>
  );
}

Snowman.defaultProps = {
  maxWrong: 6,
  images: [img0, img1, img2, img3, img4, img5, img6],
  words: ENGLISH_WORDS,
};

export default Snowman;
