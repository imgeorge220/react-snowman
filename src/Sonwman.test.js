import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Snowman from "./Snowman";

it("renders without crashing", function () {
  render(<Snowman />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<Snowman />);
  expect(asFragment()).toMatchSnapshot();
});

it("matches snapshot", function () {
  const { asFragment, queryByTestId } = render(<Snowman words={["apple"]} maxWrong={1}/>);
  
  fireEvent.click(queryByTestId('btn-b'));
  expect(asFragment()).toMatchSnapshot();
});

it("changes images on incorrect guess", function () {
  const { queryByTestId } = render(<Snowman />);

  expect(queryByTestId('img-0.png')).toBeInTheDocument();

  fireEvent.click(queryByTestId('btn-b'));
  expect(queryByTestId('img-1.png')).toBeInTheDocument();
});

it("changes letters, but not picture on correct guess", function () {
  const { queryByTestId } = render(<Snowman words={["apple"]}/>);

  expect(queryByTestId('img-0.png')).toBeInTheDocument();

  fireEvent.click(queryByTestId('btn-a'));
  expect(queryByTestId('img-0.png')).toBeInTheDocument();

  expect(queryByTestId('word')).toHaveTextContent('a');
});

it("reset button works", function () {
  const { queryByTestId } = render(<Snowman words={["apple"]}/>);

  fireEvent.click(queryByTestId('btn-a'));

  fireEvent.click(queryByTestId('reset-btn'));
  expect(queryByTestId('word')).not.toHaveTextContent('a');
});

