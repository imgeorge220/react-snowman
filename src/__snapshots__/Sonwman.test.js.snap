// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Snowman"
  >
    <img
      alt=""
      data-testid="img-0.png"
      src="0.png"
    />
    <div>
      <p>
        Number of Guesses Left: 6
      </p>
      <p
        class="Snowman-word"
        data-testid="word"
      >
        ______
      </p>
      <p>
        <button
          data-testid="btn-a"
          value="a"
        >
          a
        </button>
        <button
          data-testid="btn-b"
          value="b"
        >
          b
        </button>
        <button
          data-testid="btn-c"
          value="c"
        >
          c
        </button>
        <button
          data-testid="btn-d"
          value="d"
        >
          d
        </button>
        <button
          data-testid="btn-e"
          value="e"
        >
          e
        </button>
        <button
          data-testid="btn-f"
          value="f"
        >
          f
        </button>
        <button
          data-testid="btn-g"
          value="g"
        >
          g
        </button>
        <button
          data-testid="btn-h"
          value="h"
        >
          h
        </button>
        <button
          data-testid="btn-i"
          value="i"
        >
          i
        </button>
        <button
          data-testid="btn-j"
          value="j"
        >
          j
        </button>
        <button
          data-testid="btn-k"
          value="k"
        >
          k
        </button>
        <button
          data-testid="btn-l"
          value="l"
        >
          l
        </button>
        <button
          data-testid="btn-m"
          value="m"
        >
          m
        </button>
        <button
          data-testid="btn-n"
          value="n"
        >
          n
        </button>
        <button
          data-testid="btn-o"
          value="o"
        >
          o
        </button>
        <button
          data-testid="btn-p"
          value="p"
        >
          p
        </button>
        <button
          data-testid="btn-q"
          value="q"
        >
          q
        </button>
        <button
          data-testid="btn-r"
          value="r"
        >
          r
        </button>
        <button
          data-testid="btn-s"
          value="s"
        >
          s
        </button>
        <button
          data-testid="btn-t"
          value="t"
        >
          t
        </button>
        <button
          data-testid="btn-u"
          value="u"
        >
          u
        </button>
        <button
          data-testid="btn-v"
          value="v"
        >
          v
        </button>
        <button
          data-testid="btn-w"
          value="w"
        >
          w
        </button>
        <button
          data-testid="btn-x"
          value="x"
        >
          x
        </button>
        <button
          data-testid="btn-y"
          value="y"
        >
          y
        </button>
        <button
          data-testid="btn-z"
          value="z"
        >
          z
        </button>
      </p>
    </div>
    <button
      data-testid="reset-btn"
    >
      RESET
    </button>
  </div>
</DocumentFragment>
`;

exports[`matches snapshot 2`] = `
<DocumentFragment>
  <div
    class="Snowman"
  >
    <img
      alt=""
      data-testid="img-1.png"
      src="1.png"
    />
    <div>
      <p>
        GAME OVER!
      </p>
      <p>
        Correct word was: apple
      </p>
    </div>
    <button
      data-testid="reset-btn"
    >
      RESET
    </button>
  </div>
</DocumentFragment>
`;
